### PR TITLE
HAI-558/special transport routes

### PR DESCRIPTION
**Pull request by petri-timlin**
### Description
Added new customer data "special transport routes" including all needed phases (fetch, process, automation, config, readme).

### Jira Issue:
https://helsinkisolutionoffice.atlassian.net/browse/HAI-558

### Type of change

- [ ]  Bug fix
- [x]  New feature
- [ ] Other

### Instructions for testing
Instructions can be found from: automation/README.md:

Give following in folder /automation:
`docker build -t haitaton-gis-automation -f ./Dockerfile ..`
This wil build haitaton-gis-automation image.

Run image
Give following:
`docker run --rm --network host --env-file haitaton.env haitaton-gis-automation special_transport_routes`

Remember set all these env variables:

```
HAITATON_USER=
HAITATON_PASSWORD=
HAITATON_HOST=
HAITATON_PORT=
HAITATON_DATABASE=
HELSINKI_EXTRANET_USERNAME=
HELSINKI_EXTRANET_PASSWORD=
```
After this special_transport_routes data is written to database into the table `tormays_special_transport_routes_polys` 
and there should be following fields:

```
fid
mitlev
mitpit
mitkor
muuntaja
tuleva
varareitti
vaylatyyp2
geom
```
Data amount is about 360 rows and most of the data is clipped by ylre katuelueet areas. It might by hard to check but if you run also:
`docker run --rm --network host --env-file haitaton.env haitaton-gis-automation ylre_katualueet`
and then compare tormays_special_transport_routes_polys and tormays_ylre_classes_polys to each other.

In local environment there might be situation that table tormays_special_transport_routes_polys is missing but after docker run table should exists.

### CODE REVIEW

File: process/modules/special_transport_routes.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
+import geopandas as gpd
+import pandas as pd
+import numpy as np
+from sqlalchemy import create_engine, text
+#from shapely.validation import explain_validity
```
Comment (umeetiusbaar) [inline process/modules/special_transport_routes.py]: Can be removed?

File: process/modules/special_transport_routes.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
+import geopandas as gpd
+import pandas as pd
+import numpy as np
```
Comment (umeetiusbaar) [inline process/modules/special_transport_routes.py]: Not used?

File: process/modules/special_transport_routes.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+from sqlalchemy import create_engine, text
+#from shapely.validation import explain_validity
+from os import path
+from shapely.validation import make_valid
+from shapely.geometry import MultiPolygon, Polygon, GeometryCollection
```
Comment (umeetiusbaar) [inline process/modules/special_transport_routes.py]: Not used?

File: process/modules/special_transport_routes.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+
+        self._lines = gpd.read_file(file_name)
+        self._lines["vaylatyyp2"] = self._lines["vaylatyyp2"].str.replace("\n", "")
+
+        # Drop one invalid data line
+        self._lines = self._lines.drop(self._lines[(self._lines["id"] == 1500) | ((self._lines["nimi_1"] == "Metsälä-Etelä-Oulunkylä") & (self._lines["vaylatyyp2"] == "Kevyen liikenteen väylä"))].index)
+
+        self._orig = self._lines
+
+    def _keep_rows_base_on_hierarchy_list(self, types, shapes: gpd.GeoDataFrame) -> gpd.GeoDataFrame:
```
Comment (umeetiusbaar) [inline process/modules/special_transport_routes.py]: Can be static?

File: process/modules/special_transport_routes.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+        retval = shapes.copy()
+        mask = retval["hierarkia"].isin(types)
+        retval = retval[mask]
+        return retval
+
+    def _drop_not_used_classes_base_on_main_and_sub_types(self, main_and_sub_types, shapes: gpd.GeoDataFrame) -> gpd.GeoDataFrame:
```
Comment (umeetiusbaar) [inline process/modules/special_transport_routes.py]: Can be static?

File: process/modules/special_transport_routes.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+                        (retval["paatyyppi"] == main_type)
+                        & (retval["alatyyppi"] == sub_type)
+                    ].index,
+                    axis=0,
+                    inplace=True,
+                )
+
+        return retval
+
+    def _drop_unnecessary_columns(self, columns_to_drop: list[str], shapes: gpd.GeoDataFrame) -> gpd.GeoDataFrame:
```
Comment (umeetiusbaar) [inline process/modules/special_transport_routes.py:79]: Can be static?

File: process/modules/special_transport_routes.py [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+        attrs = ["ylre_class", "kadun_nimi"]
+        ylre_katualueet_dissolved = self._ylre_katualueet.dissolve(by=attrs)
+        ylre_katualueet_dissolved["geometry"] = ylre_katualueet_dissolved.buffer(15)
+        joined_result = gpd.sjoin(lines, ylre_katualueet_dissolved, predicate='within')
+
+        retval = lines.merge(joined_result[['gml_id','ylre_class','kadun_nimi']], how='left', left_on='gml_id', right_on='gml_id')
+
+        return retval
+
+    def _makeValid(self, data, geom_column):
```
Comment (umeetiusbaar) [inline process/modules/special_transport_routes.py]: Can be static? Also, function names should be lowercase.

File: process/modules/special_transport_routes.py [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+                connection,
+                "public",
+                if_exists="replace",
+                index=True,
+                index_label="fid",
+                )
+
+    def save_to_file(self):
+        """Save processing results to file."""
+        # Central transport routes as debug material
```
Comment (umeetiusbaar) [inline process/modules/special_transport_routes.py]: "Special" instead of "Central"?

**Review by umeetiusbaar — COMMENTED** [review]
Tested and it works described.

File: process/modules/special_transport_routes.py (hunk 2 above)
Comment (petri-timlin) [reply process/modules/special_transport_routes.py]: Removed

File: process/modules/special_transport_routes.py (hunk 1 above)
Comment (petri-timlin) [reply process/modules/special_transport_routes.py]: Removed

File: process/modules/special_transport_routes.py (hunk 3 above)
Comment (petri-timlin) [reply process/modules/special_transport_routes.py]: Removed

File: process/modules/special_transport_routes.py (hunk 4 above)
Comment (petri-timlin) [reply process/modules/special_transport_routes.py]: Removed

File: process/modules/special_transport_routes.py (hunk 6 above)
Comment (petri-timlin) [reply process/modules/special_transport_routes.py:79]: Removed

File: process/modules/special_transport_routes.py (hunk 7 above)
Comment (petri-timlin) [reply process/modules/special_transport_routes.py]: Changed function name. Function is used in process.

File: process/modules/special_transport_routes.py (hunk 8 above)
Comment (petri-timlin) [reply process/modules/special_transport_routes.py]: Changed

**Review by umeetiusbaar — APPROVED** [review]

**Pull request merged**